evm/state.py

```diff
@@ -0,0 +1,197 @@
+
+    def apply_computation(self, message):
+        """
+        Perform the computation that would be triggered by the VM message.
+        """
+        with Computation(self, message) as computation:
+            # Early exit on pre-compiles
+            if message.code_address in self.precompiles:
+                self.precompiles[message.code_address](computation)
+                return computation
+
+            for opcode in computation.code:
+                opcode_fn = self.get_opcode_fn(self.opcodes, opcode)
+
+                computation.logger.trace(
+                    "OPCODE: 0x%x (%s) | pc: %s",
+                    opcode,
+                    opcode_fn.mnemonic,
+                    max(0, computation.code.pc - 1),
+                )
+
+                try:
+                    opcode_fn(computation=computation)
+                except Halt:
+                    break
+        return computation
+
+    #
+    # Logging
+    #
+    @property
+    def logger(self):
+        return logging.getLogger('evm.vm.state.{0}'.format(self.__class__.__name__))
+
+    #
+    # Opcode API
+    #
+    def get_opcode_fn(self, opcodes, opcode):
+        try:
+            return opcodes[opcode]
+        except KeyError:
+            return InvalidOpcode(opcode)
+
+    #
+    # Block Object Properties (in opcodes)
+    #
+    @property
+    def blockhash(self):
+        return self.block_header.hash
+
+    @property
+    def coinbase(self):
+        return self.block_header.coinbase
+
+    @property
+    def timestamp(self):
+        return self.block_header.timestamp
+
+    @property
+    def number(self):
+        return self.block_header.block_number
+
+    @property
+    def difficulty(self):
+        return self.block_header.difficulty
+
+    @property
+    def gaslimit(self):
+        return self.block_header.gas_limit
+
+    #
+    # state_db
+    #
+    @contextmanager
+    def state_db(self, read_only=False):
+        state = self.chaindb.get_state_db(self.block_header.state_root, read_only)
+        yield state
+
+        if read_only:
+            # This acts as a secondary check that no mutation took place for
+            # read_only databases.
+            assert state.root_hash == self.block_header.state_root
+        elif self.block_header.state_root != state.root_hash:
+            self.block_header.state_root = state.root_hash
+
+        # remove the reference to the underlying `db` object to ensure that no
+        # further modifications can occur using the `State` object after
+        # leaving the context.
+        state.db = None
+        state._trie = None
+
+    #
+    # Snapshot and Revert
+    #
+    def snapshot(self):
+        """
+        Perform a full snapshot of the current state.
+
+        Snapshots are a combination of the state_root at the time of the
+        snapshot and the checkpoint_id returned from the journaled DB.
+        """
+        return (self.block_header.state_root, self.chaindb.snapshot())
```

Comment:
Idle musing/thought

Might be a nice API if snapshotting returned another `State` object.  This gets confusing/complex with the underlying `chaindb` which is *inherently* mutable.

Just something to think on.